St form styling

AnOctopus opened the pull request:
Forms styling, including the changes by @tvst. My recollection of having already seen changes for displaying the button warning after script stops running was due to previously reviewing changes to the main st_form branch, so merging that in completed the outstanding request by @tvst.

AnOctopus committed:
Merge branch 'develop' into st_form-styling

AnOctopus committed:
Improve styling of Forms and Submit button

FormSubmitButton no longer changes appearance when there are pending changes,
instead info text is displayed.

AnOctopus committed:
Align form status text vertically

AnOctopus committed:
Align elements in and out of forms vertically

AnOctopus committed:
Replace form padding hack with position shifting

This is less confusing, and takes care of nearly all of the visual issues that were present, because they were caused by padding ruining everything.

Also remove the styling on the submit button, because it should actually just look like the button default.

tvst committed:
Update form styling.

tvst committed:
Fix bottom margins

tvst committed:
Change styling of st.expander

tvst committed:
Lint

tvst committed:
Remove form status text

tvst committed:
Make columns use lg spacing, to match vertical spacing between elements.

AnOctopus committed:
Merge branch 'st_form' into st_form-styling

AnOctopus committed:
Fix forms merge bugs

AnOctopus committed:
Merge remote-tracking branch 'thiago/st_form-styling' into st_form-styling

tvst committed:
Improve button error message.

tvst committed:
Set form border, padding, and margin

tvst committed:
Lint.

tvst committed:
Improve form error text and styles.

tvst committed:
Remove extra space below file uploader.

tvst committed:
Remove extra space below images (had to go from inline-flex to flex, and that required a chain reaction of changes)

tvst committed:
Make st.columns horizontal gap match vertical gap between normal elements.

tvst committed:
Make st.columns have the same bottom margin as normal elements.

tvst committed:
Fix space between st.exception body and traceback.

AnOctopus committed:
Merge branch 'st_form' into st_form-styling

AnOctopus committed:
Fix lint and forms warning capitalization in merge

AnOctopus committed:
Update forms exception text in unit tests

AnOctopus committed:
Fix formatting in form test

AnOctopus committed:
Fix tests to work with Thiago's styling changes

AnOctopus committed:
Fix imports for js linting

AnOctopus committed:
Update snapshots

AnOctopus committed:
Update snapshots (that were somehow missed the first time)

AnOctopus committed:
Update expected button kind in form test

AnOctopus committed:
Update more snapshots

AnOctopus committed:
Update form tests

AnOctopus committed:
Update columns separation in tests

AnOctopus committed:
Update snapshots

AnOctopus committed:
Fix last columns test, remove snapshots to force regen

AnOctopus committed:
Add newly generated snapshots

AnOctopus committed:
Remove broken snapshots

AnOctopus committed:
Add back in removed snapshots

AnOctopus committed:
Remove snapshots so circleci can see it

AnOctopus committed:
Add many locally generated snapshots

AnOctopus committed:
Revert "Add many locally generated snapshots"

This reverts commit c32b4928b74f4319ec1391c4ecc9026b8f3ddbcc.

Commit:
modify snapshots to succeed

Commit:
Make tests pass

Commit:
Make tests pass

Commit:
Merge branch 'st_form' into foo

Commit:
Fixed test

AnOctopus committed:
Merge branch 'foo' of github.com:streamlit/streamlit into foo

AnOctopus committed:
Merge branch 'foo' into st_form-styling

kantuni commented on frontend/src/components/core/Block/styled-components.ts:
```diff
@@ -107,9 +117,28 @@ export const StyledBlock = styled.div<StyledBlockProps>(
+export const StyledForm = styled.div<StyledFormProps>(({ width, theme }) => {
+  return {
+    padding: "1em",
+    // TODO XXX: Merge with develop then use fadedText10 color.
```
Still need to resolve this TODO?

kantuni commented on frontend/src/components/widgets/Form/FormSubmitContent.tsx:
```diff
@@ -0,0 +1,14 @@
+import { FormSubmitButton, Props } from "./FormSubmitButton"
+import { StyledFormSubmitContent } from "./styled-components"
+
+export default function FormSubmitContent(props: Props): ReactElement {
```
Try exporting it without `default`. In that case the bottom export would just become `export { FormSubmitContent } from "./FormSubmitContent"`

kantuni commented on frontend/src/components/widgets/Form/index.tsx:
```diff
@@ -16,8 +16,10 @@
+import FormSubmitContent from "./FormSubmitContent"
 
 export { FormSubmitButton } from "./FormSubmitButton"
+export { FormSubmitContent }
```
here

kantuni commented on frontend/src/components/widgets/Form/styled-components.ts:
```diff
@@ -21,8 +21,24 @@ export interface StyledFormProps {
-export const StyledForm = styled.div<StyledFormProps>(({ width }) => ({
-  outline: "1px solid",
-  padding: "1em",
+export const StyledFormSubmitContent = styled.div(() => {
```
Nit: Just to be consistent
```js
() => ({
  display: "flex",
})
```

kantuni commented on frontend/src/hocs/withExpandable/withExpandable.tsx:
```diff
@@ -54,7 +54,8 @@ function withExpandable(
 
     const toggle = (): void => toggleExpanded(!expanded)
-    const { colors, fontWeights, spacing } = useTheme<Theme>()
+    const theme: Theme = useTheme()
```
IMHO, the previous one looked cleaner.

tconkling commented on frontend/src/components/widgets/Form/Form.tsx:
```diff
@@ -33,11 +33,20 @@ interface State {
   "\n\nThis form has no submit button, which means that user interactions will " +
   "never be sent to your Streamlit app." +
-  "\n\nTo create a submit button, use the `st.beta_submit_button()` function."
+  "\n\nTo create a submit button, use the `st.submit_button()` function." +
+  "\n\nFor more information, refer to the " +
+  "[documentation for forms](https://docs.streamlit.io/api.html#form)."
```
The API here is `st.form_submit_button()`

tconkling commented on frontend/src/components/widgets/Form/Form.tsx:
```diff
@@ -33,11 +33,20 @@ interface State {
   showMissingSubmitButtonWarning: boolean
 }
 
+/**
+ * If the form doesn't have a submit button after this many milliseconds
+ * have elapsed, we show a warning that tells them they probably want to
+ * create one.
+ */
+export const SUBMIT_BUTTON_ERROR_TIME_MS = 1500
+
```
This looks like it snuck in from a bad rebase, maybe? The timer is no longer part of the form implementation, so this should be removed.

tconkling reviewed (commented):
Huzzah! Thanks for all your work on this, @AnOctopus. I know it was a huge pain. I'll let @kantuni handle the bulk of the review, but I've left a couple minor comments here.

(Also: I haven't actually run the code, but just confirming: Submit buttons still have two different visual states, one for "form has unsubmitted changes" and one for "form has no unsubmitted changes", right?)

AnOctopus commented:
> (Also: I haven't actually run the code, but just confirming: Submit buttons still have two different visual states, one for "form has unsubmitted changes" and one for "form has no unsubmitted changes", right?)

Nope, just one now

tconkling commented:
Oh - did Thiago change his mind about that?

asaini commented:
@tconkling : We decided to not use different visual states, since users of the form could be interested in submitting the default state of the form as well. 

So we simplified to just one state.

AnOctopus committed:
Forms code review cleanup

Replace color per TODO comment
Remove now unused constant
Fix function reference in warning message
Clean up exports
Simplify ts destructuring and object returns
